src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+    /// [`PlanTrace`] even though its receiver is not `&mut self`. This quirk is
+    /// required because the tracing `Dispatch` does not have `downcast_mut` method.
+    pub fn drain_as_vec(&self) -> Vec<TraceEntry<T>>
+    where
+        T: Debug,
+    {
+        let mut entries = self.entries.lock().expect("entries shouldn't be poisoned");
+        entries.split_off(0)
+    }
+
+    /// Push a trace entry for the given `plan` to the current trace.
+    ///
+    /// This is a noop if the call is within a context without an enclosing span.
+    fn push(&self, plan: T, time: std::time::Instant) {
+        let times = self.times.lock().expect("times shouldn't be poisoned");
+        if let Some(span_start) = times.last() {
+            let mut entries = self.entries.lock().expect("entries shouldn't be poisoned");
+            entries.push(TraceEntry {
+                instant: time,
+                duration: time.duration_since(*span_start),
+                path: self.current_path(),
+                plan,
+            });
+        }
+    }
+
+    /// Helper method: get a copy of the current path.
+    fn current_path(&self) -> String {
+        let path = self.path.lock().expect("path shouldn't be poisoned");
+        path.deref().display().to_string()
+    }
+}
+
+/// Helper trait used to extract attributes of type `&'static str`.
+trait GetStr {
+    fn get_str(&self, key: &'static str) -> Option<String>;
+}
+
+impl<'a> GetStr for span::Attributes<'a> {
+    fn get_str(&self, key: &'static str) -> Option<String> {
+        let mut extract_str = ExtractStr::new(key);
+        self.record(&mut extract_str);
+        extract_str.val()
+    }
+}
+
+/// Helper struct that implements `field::Visit` and is used in the
+/// `GetStr::get_str` implementation for `span::Attributes`.
+struct ExtractStr {
+    key: &'static str,
+    val: Option<String>,
+}
+
+impl ExtractStr {
+    fn new(key: &'static str) -> Self {
+        Self { key, val: None }
+    }
+
+    fn val(self) -> Option<String> {
+        self.val
+    }
+}
+
+impl field::Visit for ExtractStr {
+    fn record_str(&mut self, field: &tracing::field::Field, value: &str) {
+        if field.name() == self.key {
+            self.val = Some(value.to_string())
+        }
+    }
+
+    fn record_debug(&mut self, _field: &tracing::field::Field, _value: &dyn std::fmt::Debug) {}
+}
+
+#[cfg(test)]
+mod test {
+    use tracing::dispatcher;
+    use tracing::instrument;
+    use tracing_subscriber::prelude::*;
+
+    use super::trace_plan;
+    use super::PlanTrace;
+
+    #[test]
+    fn test_optimizer_trace() {
+        let subscriber = tracing_subscriber::registry().with(Some(PlanTrace::<String>::new()));
+        let dispatch = dispatcher::Dispatch::new(subscriber);
+
+        dispatcher::with_default(&dispatch, || {
+            optimize();
+        });
+
+        if let Some(trace) = dispatch.downcast_ref::<PlanTrace<String>>() {
```

Comment:
nit: If the `Dispatch` API had a `downcast_mut` counterpart I would be able to define `drain_as_vec` on `&mut self`.

Reply:
This will probably never be added so doing the coordination yourself is fine!